File: services/producers/java/src/main/java/com/example/producer/Producer.java
package com.example.producer;

import com.rabbitmq.client.*;
import java.nio.charset.StandardCharsets;

public class Producer {
    private static final String EXCHANGE_NAME = "events-exchange";

    public static void main(String[] argv) throws Exception {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost("rabbitmq");
        factory.setUsername("admin");
        factory.setPassword("admin");

        try (Connection connection = factory.newConnection();
             Channel channel = connection.createChannel()) {

            // Declarar el exchange fanout y durable
            channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.FANOUT, true);

            for (int i = 0; i < 5; i++) {
                String message = String.format("{\"source\":\"producer-java\",\"message\":\"event %d\"}", i);
                channel.basicPublish(EXCHANGE_NAME, "", 
                    MessageProperties.PERSISTENT_TEXT_PLAIN,
                    message.getBytes(StandardCharsets.UTF_8));
                System.out.println("Sent: " + message);
                Thread.sleep(1000);
            }
        }
    }
}

File: services/producers/python/producer.py
import pika, json, time, os

RABBIT_HOST = os.getenv("RABBIT_HOST", "rabbitmq")
credentials = pika.PlainCredentials("admin", "admin")

while True:
    try:
        connection = pika.BlockingConnection(pika.ConnectionParameters(RABBIT_HOST, credentials=credentials))
        channel = connection.channel()
        channel.exchange_declare(exchange="events-topic-exchange", exchange_type="topic", durable=True)

        i = 0
        while True:
            msg = {"source": "producer-python", "message": f"event {i}"}
            channel.basic_publish(
                exchange="events-topic-exchange",
                routing_key="events.python",
                body=json.dumps(msg),
                properties=pika.BasicProperties(delivery_mode=2),
            )
            print("Sent:", msg)
            i += 1
            time.sleep(5)
    except Exception as e:
        print("Connection failed, retrying...", e)
        time.sleep(3)